File: reorder.ts
import * as fs from "fs";

const file = fs.readFileSync("src/pages/Movies.tsx", "utf8");

const prefix = file.slice(0, file.indexOf("export const MOVIE_PROVIDERS = ["));
const suffixIdx = file.indexOf("];\n\nexport default function Movies");
const arrayStr = file.slice(file.indexOf("export const MOVIE_PROVIDERS = [") + "export const MOVIE_PROVIDERS = [".length, suffixIdx);
const suffix = file.slice(suffixIdx);

const providers = [];
let braceCount = 0;
let current = "";

for (let i = 0; i < arrayStr.length; i++) {
    const char = arrayStr[i];
    if (char === "{") {
        braceCount++;
    }
    if (braceCount > 0) {
        current += char;
    }
    if (char === "}") {
        braceCount--;
        if (braceCount === 0) {
            providers.push(current);
            current = "";
        }
    }
}

const order = [
  "4khdhub.one",
  "uhdmovies.food",
  "new3.moviesdrives.my",
  "vegamovies.mq",
  "kmmovies.lol",
  "multishows.top",
  "katmoviehd.cymru",
  "top.xdmovies.wtf",
  "new.cloudmoviez.shop",
  "tamiltvtoons.site",
  "movienestbd.pics",
  "v2.olamovies.mov",
  "ddlbase.com",
  "zinkmovies.today",
  "cinemalux.wiki",
  "cinefreak.nl",

  "mlsbd.co",
  "joya9tv1.com",
  "cinedoze.tv",
  "southfreak.ink",
  "moviebaaz.cfd",
  "moviedokan.co",
  "moviedrivebd.com",
  "freedrivemovie.cfd",

  "fojik.site",
  "newhdmovie2.top",
  "mlfbd.best",
  "new.bollyflix.gd",
  "a.privatemoviez.surf",
  "new1.hdhub4u.cl",
  "go.india4movies.net",
  "ssrmovies.taxi",
  "downloadhub.lat",
  "moviesleech.link",
  "moviesmod.farm",
  "hdmovieverse.xyz",
  "1tamilmv.futbol",
  "allmovieshub.gives",
  "world4ufree.tw",
  "www.thenextplanet.living",
  "mkvbase.site",
  "moviedbhub.com"
];

const getDomain = (p) => {
    const match = p.match(/domain:\s*'([^']+)'/);
    return match ? match[1] : "";
};

const uniqueProvidersMap = new Map();
providers.forEach(p => {
    const domain = getDomain(p);
    if (!uniqueProvidersMap.has(domain)) {
        uniqueProvidersMap.set(domain, p);
    } else {
        const existing = uniqueProvidersMap.get(domain);
        if (p.includes("logo: '/logos/") && !existing.includes("logo: '/logos/")) {
            uniqueProvidersMap.set(domain, p);
        }
    }
});

const uniqueProviders = Array.from(uniqueProvidersMap.values());

const sorted = [];
order.forEach(domain => {
    const p = uniqueProviders.find(p => getDomain(p) === domain);
    if (p) {
        sorted.push(p);
    }
});

uniqueProviders.forEach(p => {
    const domain = getDomain(p);
    if (!order.includes(domain)) {
        sorted.push(p);
    }
});

const newFile = prefix + "export const MOVIE_PROVIDERS = [\n  " + sorted.join(",\n  ") + "\n" + suffix;
fs.writeFileSync("src/pages/Movies.tsx", newFile);
console.log("Done");

File: reorder.py
import re

with open('src/data/searchResources.ts', 'r') as f:
    text = f.read()

# We can find all blocks `{ ... }`
# A block starts with `{` at the beginning of a line with indentation `  {`
# and ends with `  }` or `  },`
# So we can match them using `r'  \{\n(?:.*?\n)*?  \},?'`

blocks = re.findall(r'  \{\n(?:.*?\n)*?  \},?', text)

# Now we categorize them
xd_domain = "top.xdmovies.wtf"
move_domains = [
    "cinefreak.nl",
    "mlsbd.co",
    "joya9tv1.com",
    "cinedoze.tv",
    "southfreak.fyi",
    "moviebaaz.cfd",
    "moviedokan.co",
    "moviedrivebd.com",
    "freedrivemovie.cfd",
    "fojik.site",
    "movienestbd.pics",
    "go.india4movies.net",
    "a.privatemoviez.surf",
    "new1.hdhub4u.cl"
]

def get_domain(block):
    m = re.search(r'domain:\s*["\']([^"\']+)["\']', block)
    return m.group(1).strip() if m else None

xd_block = None
move_blocks_dict = {}
other_blocks = []

for b in blocks:
    d = get_domain(b)
    if d == xd_domain:
        xd_block = b
    elif d in move_domains:
        move_blocks_dict[d] = b
    else:
        other_blocks.append(b)

# Construct new array contents
new_blocks = []
xd_index = -1

# We will put other_blocks in their original order, 
# and insert the move_domains right after xd_domain.
for i, b in enumerate(other_blocks):
    # wait, xd_block isn't in other_blocks.
    pass

# Actually, let's keep the order of everything that is not in move_domains.
# But where does xd_block go? Where it currently is among the remaining elements.
new_blocks_base = []
for b in blocks:
    d = get_domain(b)
    if d not in move_domains:
        new_blocks_base.append(b)

# Now find where xd_block is in new_blocks_base
final_blocks = []
for b in new_blocks_base:
    final_blocks.append(b)
    if get_domain(b) == xd_domain:
        # Append the move blocks in requested order
        for md in move_domains:
            if md in move_blocks_dict:
                # Ensure it has a comma if it's not the last one, but we can fix commas later.
                final_blocks.append(move_blocks_dict[md])

# Fix commas: all except last should end with `,`
for i in range(len(final_blocks)):
    final_blocks[i] = final_blocks[i].rstrip(",\n")
    if i < len(final_blocks) - 1:
        final_blocks[i] += ",\n"
    else:
        final_blocks[i] += "\n"

# Replace in file
# We'll just split text before the first block and after the last block
first_block_idx = text.find(blocks[0])
last_block_idx = text.find(blocks[-1]) + len(blocks[-1])

new_text = text[:first_block_idx] + "".join(final_blocks) + text[last_block_idx:]

with open('src/data/searchResources.ts', 'w') as f:
    f.write(new_text)

